subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -111,14 +129,22 @@ protected void handleIterable(Object node, BiConsumer<String, Object> handler) {
         }
 
         @Override
-        protected Object getChild(Object parent, PropertyMetadata property) {
+        protected Optional<Object> getChild(Object parent, PropertyMetadata property) {
             try {
-                return property.getGetterMethod().invoke(parent);
+                return Optional.ofNullable(property.getGetterMethod().invoke(parent));
             } catch (Exception e) {
                 // TODO Handle this
                 throw new RuntimeException(e);
             }
         }
+
+        @Override
+        protected Set<Object> newIdentitySetOf(Set<Object> initialSet, Object newElement) {
+            Set<Object> set = Sets.newIdentityHashSet();
+            set.addAll(initialSet);
+            set.add(newElement);
+            return Collections.unmodifiableSet(set);
```

Comment:
When we throw for a cycle, it would be good to show an error message with a list of the nodes in the cycle. In the message we need to list the nodes in order, and get the names of the properties themselves. The currently used `IdentityHashMap` does not provide either of these.

We could use a `LinkedHashSet<NamedValue>` where `NamedValue` would only take into account the identity of the value itself for equality.